core/src/main/java/com/yahoo/oak/EntrySet.java

```diff
@@ -660,7 +656,7 @@ void writeValueStart(ThreadContext ctx, V value, boolean writeForMove) {
         // The allocated slice is actually the thread's ByteBuffer moved to point to the newly
```

Comment:
It is important to mention that having a small block size (8MB vs 256MB) also limits the ability to write big values. The value size can be maximum the size of block. Currently, trial to allocate and write a value bigger than a block will result in out-of-bound exception. I suggest to add here a more clear exception, printing explanation that values bigger than block can not be allocated/written. After all it was a user decision to define small blocks...

Reply:
Please note that this problem does not relate to these code modifications, but we can address it now if you prefer.

**Background:**
Even with larger block sizes, the maximal value length was 8MB and the maximal key length was 64KB due to the reference encoding limitations.

With a small block size, trying to allocate a key/value larger than the block size will actually issue an `OakOutOfMemoryException` (see NativeMemoryAllocator:114-116), which might make sense albeit we can add there an additional message in the exception that explains the error.

On the other hand, trying to allocate a large value when using a larger block size will not be detected at all.
I somewhat addressed that in the `ReferenceCodec` class by adding an assertion there, but it only applied for testing.
Production code will not get ANY error until it will try to access the data and then it will simply get an out-of-bound exception as you mentioned. But that is also true before this PR.

My suggestions:
* Replacing the assertion in `ReferenceCodec` with an actual production-time validation that may raise `IllegalArgumentException` (may have some minor performance implications).
* Replacing the exception in NativeMemoryAllocator:115 also with `IllegalArgumentException`.

Let me know if you prefer to add these modifications now or discuss them further to be included in a different PR.

Reply:
I think here (when we first get the requested value size) is the place to check both for the block size limitation and for length encoding limitation. The information about what the limitations are, can be taken from ReferenceCodec. 

And although the length encoding limitation was already there, the block size limitation is harder now. When the possibility is given to the user to set the size of the block, it can be set to less than 8MB, thus this limit can be the first to fire.